solidity/contracts/ECDSARewards.sol

```diff
@@ -158,16 +173,41 @@ contract ECDSARewards is Rewards {
         return factory.getKeepOpenedTimestamp(toAddress(_keep)) != 0;
     }
 
+    // Get the members of the specified keep,
+    // and distribute the reward amount between them.
+    // The reward isn't paid out immediately,
+    // but is instead kept in the reward contract
+    // until each operator individually requests to withdraw the rewards.
     function _distributeReward(bytes32 _keep, uint256 amount)
         internal
         isAddress(_keep)
     {
-        token.approve(toAddress(_keep), amount);
+        address[] memory members = BondedECDSAKeep(toAddress(_keep)).getMembers();
+        uint256 interval = intervalOf(_getCreationTime(_keep));
 
-        BondedECDSAKeep(toAddress(_keep)).distributeERC20Reward(
-            address(token),
-            amount
-        );
+        uint256 memberCount = members.length;
+        uint256 dividend = amount.div(memberCount);
+        uint256 remainder = amount.mod(memberCount);
+
+        uint256[] memory allocations = new uint256[](memberCount);
+
+        for (uint256 i = 0; i < memberCount - 1; i++) {
+            allocations[i] = dividend;
+        }
+        allocations[memberCount - 1] = dividend.add(remainder);
+
+        for (uint256 i = 0; i < memberCount; i++) {
+            address member = members[i];
+            uint256 addedAllocation = allocations[i];
+            uint256 prevAllocated = allocatedRewards[member][interval];
+            uint256 newAllocation = prevAllocated.add(addedAllocation);
+            if (newAllocation > operatorRewardCap) {
+                uint256 deallocatedAmount = newAllocation.sub(operatorRewardCap);
+                newAllocation = operatorRewardCap;
+                unallocatedRewards = unallocatedRewards.add(deallocatedAmount);
```

Comment:
We may have a little problem with upgrades. When finalizing the upgrade, we are allocating all past intervals and then, tokens are moved to the new contract. Rewards for past intervals are distributed later and we may find ourselves in a situtation when there is a lot of unallocated rewards on the previous rewards contract and there is an incetive to stay there.

What do you think? One option could be to disable any allocations after finalizing the upgrade and having a way to send all unallocated rewards to the new contract.

Reply:
Good catch, the unallocated rewards line should be using `sendToNewRewards` instead, which takes care of this correctly.

Reply:
Derp, I was looking at the wrong version of the rewards upgrade. We should do this yes:

> One option could be to disable any allocations after finalizing the upgrade and having a way to send all unallocated rewards to the new contract.

Reply:
Addressed in https://github.com/keep-network/keep-core/pull/2136 and later in 3616caf58f750839b45bf9a268d12912ef9cba61.